ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+        ProviderMessaging_TrySendRequestAndWaitForResponse,
+        root,
+        messageType,
+        vnode,
+        vnodeFsidInode,
+        vnodePath,
+        pid,
+        procname,
+        kauthResult,
+        kauthError);
+    
+    return true;
+}
+
+static void SetFileXattrData(shared_ptr<vnode> vnode)
+{
+    PrjFSFileXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    vnode->xattrs.insert(make_pair(PrjFSFileXAttrName, rootXattrData));
+}
+
+@interface HandleVnodeOperationTests : XCTestCase
+@end
+
+@implementation HandleVnodeOperationTests
+{
+    vfs_context_t context;
+    const char* repoPath;
+    const char* filePath;
+    const char* dirPath;
+    PrjFSProviderUserClient dummyClient;
+    pid_t dummyClientPid;
+    shared_ptr<mount> testMount;
+    shared_ptr<vnode> repoRootVnode;
+    shared_ptr<vnode> testFileVnode;
+    shared_ptr<vnode> testDirVnode;
+}
+
+- (void) setUp
+{
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+    context = vfs_context_create(NULL);
+    dummyClientPid = 100;
+
+    // Create Vnode Tree
+    repoPath = "/Users/test/code/Repo";
+    filePath = "/Users/test/code/Repo/file";
+    dirPath = "/Users/test/code/Repo/dir";
+    testMount = mount::Create();
+    repoRootVnode = testMount->CreateVnodeTree(repoPath, VDIR);
+    testFileVnode = testMount->CreateVnodeTree(filePath);
+    testDirVnode = testMount->CreateVnodeTree(dirPath, VDIR);
+
+    // Register provider for the repository path (Simulate a mount)
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&dummyClient, dummyClientPid, repoPath);
+    XCTAssertEqual(result.error, 0);
+    vnode_put(s_virtualizationRoots[result.root].rootVNode);
+}
+
+- (void) tearDown
+{
+    testMount.reset();
+    repoRootVnode.reset();
+    testFileVnode.reset();
+    testDirVnode.reset();
+    VirtualizationRoots_Cleanup();
+    vfs_context_rele(context);
+    MockVnodes_CheckAndClear();
+    MockCalls::Clear();
+}
+
+- (void) testEmptyFileHydrates {
+    testFileVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    
+    const int actionCount = 8;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        KAUTH_VNODE_WRITE_ATTRIBUTES,
+        KAUTH_VNODE_READ_EXTATTRIBUTES,
+        KAUTH_VNODE_WRITE_EXTATTRIBUTES,
+        KAUTH_VNODE_READ_DATA,
+        KAUTH_VNODE_WRITE_DATA,
+        KAUTH_VNODE_EXECUTE,
+        KAUTH_VNODE_DELETE,
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testFileVnode.get()),
+            0,
+            0);
+        XCTAssertTrue(
```

Comment:
Also validate that `KAUTH_VNODE_` actions that should not hydrate do not trigger `MessageType_KtoU_HydrateFile`? (e.g. `KAUTH_VNODE_READ_SECURITY`)